File: training/comun/suavizado.py
"""
Suavizado temporal de landmarks con el filtro One Euro.

La Tasks API de MediaPipe (la vigente) no aplica ningun filtro a los landmarks,
asi que estos tiemblan fotograma a fotograma por el ruido de la deteccion. Este
modulo implementa el filtro One Euro, disenado justo para senales interactivas:
suaviza fuerte cuando la senal se mueve lento (la mano quieta en una letra
estatica queda firme) y suaviza poco cuando se mueve rapido (la J, la Z o la LL
no se retrasan). Un promedio movil simple no sirve porque agregaria retraso a las
letras con movimiento.

El mismo filtro se aplica en dos lugares que DEBEN coincidir:
  - Preprocesamiento: sobre cada secuencia capturada, antes de la escala y las
    ventanas, para que el Modelo A entrene con landmarks ya suaves.
  - Inferencia (demo y app): en linea, fotograma a fotograma.

Se aplica sobre el vector YA TRASLADADO (muneca en el origen), igual en ambos
lados. Las coordenadas que valen exactamente cero (muneca en el origen y mitad de
una mano ausente) se preservan en cero, para no romper el contrato que usa la
normalizacion por escala.

Referencia: G. Casiez, N. Roussel, D. Vogel, "1 Euro Filter: A Simple Speed-based
Low-pass Filter for Noisy Input in Interactive Systems", CHI 2012.
"""

import numpy as np


def _alpha(cutoff: float, dt: float):
    """Coeficiente del filtro pasa-bajos para una frecuencia de corte dada."""
    tau = 1.0 / (2.0 * np.pi * cutoff)
    return 1.0 / (1.0 + tau / dt)


class FiltroUnEuro:
    """Filtro One Euro vectorizado para un vector de coordenadas.

    Mantiene estado entre llamadas, asi que sirve tanto en linea (un fotograma
    por llamada en la demo) como por lotes (recorriendo una secuencia). Cada
    coordenada se filtra de forma independiente con su propia frecuencia de corte
    adaptativa.
    """

    def __init__(self, fps: float = 30.0, min_cutoff: float = 1.0,
                 beta: float = 1.0, d_cutoff: float = 1.0):
        self.dt = 1.0 / float(fps)
        self.min_cutoff = float(min_cutoff)
        self.beta = float(beta)
        self.d_cutoff = float(d_cutoff)
        self._x_prev = None
        self._dx_prev = None

    def reiniciar(self) -> None:
        """Olvida el estado. Llamar cuando la mano desaparece, para que al volver
        el filtro arranque limpio y no interpole un salto."""
        self._x_prev = None
        self._dx_prev = None

    def filtrar(self, x, preservar_ceros: bool = True):
        """Filtra un vector de coordenadas y devuelve la version suavizada.

        Si `preservar_ceros`, las coordenadas que entran en exactamente cero salen
        en cero (muneca en el origen y mano ausente).
        """
        x = np.asarray(x, dtype=np.float64)
        if self._x_prev is None:
            self._x_prev = x.copy()
            self._dx_prev = np.zeros_like(x)
            salida = x.copy()
        else:
            dx = (x - self._x_prev) / self.dt
            a_d = _alpha(self.d_cutoff, self.dt)
            dx_hat = a_d * dx + (1.0 - a_d) * self._dx_prev
            cutoff = self.min_cutoff + self.beta * np.abs(dx_hat)
            a = _alpha(cutoff, self.dt)  # un alpha por coordenada
            salida = a * x + (1.0 - a) * self._x_prev
            self._x_prev = salida
            self._dx_prev = dx_hat
        if preservar_ceros:
            salida = np.where(x == 0.0, 0.0, salida)
        return salida


def suavizar_secuencia(seq, fps: float = 30.0, min_cutoff: float = 1.0,
                       beta: float = 1.0, d_cutoff: float = 1.0):
    """Suaviza una secuencia (T, D) recorriendola de forma causal, como en vivo.

    Usa un filtro nuevo por secuencia, para que cada toma se suavice de forma
    independiente (igual que cuando la mano aparece en la demo). Preserva los
    ceros. Devuelve un array float32 de la misma forma.
    """
    seq = np.asarray(seq, dtype=np.float32)
    filtro = FiltroUnEuro(fps, min_cutoff, beta, d_cutoff)
    salida = np.empty_like(seq)
    for i in range(len(seq)):
        salida[i] = filtro.filtrar(seq[i])
    return salida.astype(np.float32)

File: app/lib/core/suavizado.dart
import 'dart:math' as math;

/// CONTRATO CRITICO: replica exactamente `FiltroUnEuro` de
/// [training/comun/suavizado.py]. Suaviza los landmarks temblorosos que entrega
/// MediaPipe (la Tasks API no filtra). Se aplica sobre el vector YA TRASLADADO
/// (muneca en el origen), igual que en el preprocesamiento y en la demo, antes de
/// la escala. Sin este filtro la app veria landmarks con temblor que el modelo
/// nunca vio al entrenar.
///
/// Suaviza fuerte cuando la mano se mueve lento (letra estatica firme) y poco
/// cuando se mueve rapido (la J o la LL no se retrasan): un promedio movil no
/// sirve porque agregaria retraso a las letras con movimiento.
double _alpha(double cutoff, double dt) {
  final tau = 1.0 / (2.0 * math.pi * cutoff);
  return 1.0 / (1.0 + tau / dt);
}

class FiltroUnEuro {
  final double dt;
  final double minCutoff;
  final double beta;
  final double dCutoff;

  List<double>? _xPrev;
  List<double>? _dxPrev;

  FiltroUnEuro({
    double fps = 30.0,
    this.minCutoff = 1.0,
    this.beta = 1.0,
    this.dCutoff = 1.0,
  }) : dt = 1.0 / fps;

  /// Olvida el estado. Llamar cuando la mano desaparece, para que al volver el
  /// filtro arranque limpio y no interpole un salto.
  void reiniciar() {
    _xPrev = null;
    _dxPrev = null;
  }

  /// Filtra un vector de coordenadas y devuelve la version suavizada.
  ///
  /// [dt] es el tiempo real (segundos) desde el fotograma anterior. Es IMPORTANTE
  /// pasarlo cuando la tasa de fotogramas no es la nominal (en el telefono corre
  /// mas lento que 30 fps): con el dt correcto el suavizado se comporta igual que
  /// a 30 fps para un mismo movimiento real. Si no se pasa, usa el dt nominal.
  ///
  /// Si [preservarCeros], las coordenadas que entran en exactamente cero salen en
  /// cero (muneca en el origen y mitad de una mano ausente).
  List<double> filtrar(List<double> x, {double? dt, bool preservarCeros = true}) {
    final paso = dt ?? this.dt;
    final n = x.length;
    if (_xPrev == null) {
      _xPrev = List<double>.from(x);
      _dxPrev = List<double>.filled(n, 0.0);
      return List<double>.from(x); // primera llamada: pasa igual (los ceros ya son cero)
    }

    final xp = _xPrev!;
    final dxp = _dxPrev!;
    final aD = _alpha(dCutoff, paso);
    final salida = List<double>.filled(n, 0.0);
    for (var i = 0; i < n; i++) {
      final dx = (x[i] - xp[i]) / paso;
      final dxHat = aD * dx + (1.0 - aD) * dxp[i];
      final cutoff = minCutoff + beta * dxHat.abs();
      final a = _alpha(cutoff, paso); // un alpha por coordenada
      final suave = a * x[i] + (1.0 - a) * xp[i];
      // El estado guarda el valor suave crudo; la salida preserva los ceros.
      xp[i] = suave;
      dxp[i] = dxHat;
      salida[i] = (preservarCeros && x[i] == 0.0) ? 0.0 : suave;
    }
    return salida;
  }
}
